kse/src/org/kse/gui/actions/SignJwtAction.java

```diff
@@ -0,0 +1,167 @@
+
+/**
+ * Action to create a JWT (JSON Web Token)
+ *
+ */
+public class SignJwtAction extends KeyStoreExplorerAction {
+
+	private static final long serialVersionUID = 1L;
+
+	/**
+	 * Construct action.
+	 * 
+	 * @param kseFrame KeyStore Explorer frame
+	 */
+	public SignJwtAction(KseFrame kseFrame) {
+		super(kseFrame);
+		putValue(LONG_DESCRIPTION, res.getString("SignJwtAction.statusbar"));
+		putValue(NAME, res.getString("SignJwtAction.text"));
+		putValue(SHORT_DESCRIPTION, res.getString("SignJwtAction.tooltip"));
+		putValue(SMALL_ICON,
+				new ImageIcon(Toolkit.getDefaultToolkit().createImage(getClass().getResource("images/signcrl.png"))));
+	}
+
+	@Override
+	protected void doAction() {
+		try {
+			KeyStoreHistory history = kseFrame.getActiveKeyStoreHistory();
+			KeyStoreState currentState = history.getCurrentState();
+
+			String alias = kseFrame.getSelectedEntryAlias();
+
+			Password password = getEntryPassword(alias, currentState);
+
+			if (password == null) {
+				return;
+			}
+			KeyStore keyStore = currentState.getKeyStore();
+
+			Provider provider = history.getExplicitProvider();
+			PrivateKey privateKey = (PrivateKey) keyStore.getKey(alias, password.toCharArray());
+			KeyPairType keyPairType = KeyPairUtil.getKeyPairType(privateKey);
+
+			DSignJwt dSignJwt = new DSignJwt(frame, keyPairType, privateKey);
+			dSignJwt.setLocationRelativeTo(frame);
+			dSignJwt.setVisible(true);
+			if (dSignJwt.isOk()) {
+				String encodedJWT = signJwt(dSignJwt, privateKey, provider);
+				DViewJwt dialog = new DViewJwt(frame, encodedJWT);
+				dialog.setLocationRelativeTo(frame);
+				dialog.setVisible(true);
+			}
+		} catch (Exception ex) {
+			DError.displayError(frame, ex);
+		}
+	}
+
+	private String signJwt(DSignJwt dSignJwt, PrivateKey privateKey, Provider provider)
+			throws Exception {
+
+		Curve curve = null;
+		JWSAlgorithm signatureAlgorithm = null;
+		SignatureType signatureType = dSignJwt.getSignatureType();
+		switch (signatureType) {
+		case SHA256_RSA:
+			signatureAlgorithm = JWSAlgorithm.RS256;
+			break;
+		case SHA384_RSA:
+			signatureAlgorithm = JWSAlgorithm.RS384;
+			break;
+		case SHA512_RSA:
+			signatureAlgorithm = JWSAlgorithm.RS512;
+			break;
+		case SHA256WITHRSAANDMGF1:
+			signatureAlgorithm = JWSAlgorithm.PS256;
+			break;
+		case SHA384WITHRSAANDMGF1:
+			signatureAlgorithm = JWSAlgorithm.PS384;
+			break;
+		case SHA512WITHRSAANDMGF1:
+			signatureAlgorithm = JWSAlgorithm.PS512;
+			break;
+		case SHA256_ECDSA:
+			signatureAlgorithm = JWSAlgorithm.ES256;
+			curve = Curve.P_256;
+			break;
+		case SHA384_ECDSA:
+			signatureAlgorithm = JWSAlgorithm.ES384;
+			curve = Curve.P_384;
+			break;
+		case SHA512_ECDSA:
+			signatureAlgorithm = JWSAlgorithm.ES512;
+			curve = Curve.P_521;
+			break;
+		default:
+			break;
+		}
+		if (signatureAlgorithm == null) {
+			throw new NoSuchAlgorithmException(
+					signatureType + ": " + res.getString("SignJwtAction.signNotAvailable.message"));
+		}
```

Comment:
Hello @jgrateron (and @kaikramer),

Here is a question or a minor request:

- When signature type is not available _(for example, using SHA-1,...)_;
Instead of letting java do it and having an exception _(even if you added a message)_,
could you instead make a simple error message without making a java exception _(with java-stack, etc)_...

Example observed, with testing a Sign JWT with  `SHA-1 with RSA`:
 
![2022-01-30_Capture_JWT_SHA-1](https://user-images.githubusercontent.com/86879521/151716462-2e54400b-7aad-4d51-b59f-04f7441ba1a1.png)

```.cy
java.security.NoSuchAlgorithmException: SHA-1 with RSA: Signature not available
	at org.kse.gui.actions.SignJwtAction.signJwt(SignJwtAction.java:129)
	at org.kse.gui.actions.SignJwtAction.doAction(SignJwtAction.java:77)
	at org.kse.gui.actions.KeyStoreExplorerAction.actionPerformed(KeyStoreExplorerAction.java:93)
   ...
```

Regards.

Reply:
Hello yes of course, you can make the adjustment to generate a simple error message.